src/optimizer/rule/comparison_simplification.cpp

```diff
@@ -35,25 +35,42 @@ unique_ptr<Expression> ComparisonSimplificationRule::Apply(LogicalOperator &op,
 	}
 	if (column_ref_expr->expression_class == ExpressionClass::BOUND_CAST) {
 		//! Here we check if we can apply the expression on the constant side
+		//! We can do this if the cast itself is invertible and casting the constant is
+		//! invertible in practice.
 		auto cast_expression = (BoundCastExpression *)column_ref_expr;
 		auto target_type = cast_expression->source_type();
-		if (!BoundCastExpression::CastIsInvertible(target_type, cast_expression->return_type) ||
-		    !BoundCastExpression::CastIsInvertible(cast_expression->return_type, target_type)) {
+		if (!BoundCastExpression::CastIsInvertible(target_type, cast_expression->return_type)) {
 			return nullptr;
 		}
-		auto new_constant = constant_value.TryCastAs(target_type, true);
-		if (new_constant) {
-			auto child_expression = move(cast_expression->child);
-			auto new_constant_expr = make_unique<BoundConstantExpression>(constant_value);
-			//! We can cast, now we change our column_ref_expression from an operator cast to a column reference
-			if (column_ref_left) {
-				expr->left = move(child_expression);
-				expr->right = move(new_constant_expr);
-			} else {
-				expr->left = move(new_constant_expr);
-				expr->right = move(child_expression);
+
+		// Can we cast the constant at all?
+		string error_message;
+		Value cast_constant;
+		auto new_constant = constant_value.TryCastAs(target_type, cast_constant, &error_message, true);
+		if (!new_constant) {
+			return nullptr;
+		}
+
+		// Is the constant cast invertible?
+		if (!BoundCastExpression::CastIsInvertible(cast_expression->return_type, target_type)) {
+			// Is it actually invertible?
+			Value uncast_constant;
+			if (!cast_constant.TryCastAs(constant_value.type(), uncast_constant, &error_message, true) ||
+			    uncast_constant != constant_value) {
+				return nullptr;
```

Comment:
Maybe the invertibility check of a constant could be moved to a method in value?

Reply:
After looking again, there are too many dependencies here to make a clean break. Among other things, we need the successful first cast later on. Plus it would make `Value` depend on `BoundCastExpression`, which seems wrong.